### PR TITLE
Thread and Database Abstraction and Optimisation

### LINKED ISSUE #722: Add redis configuration support
**Is your suggestion related to a problem? Please describe in detail.**
A clear and concise description of what the problem is.
I tried to set up skycrypt, but I only have a remote redis instance to use.
Currently there is no option to configure the redis hostname.

**Describe your suggestion**
Add an option in the `credentials.json` file to add this configuration option

**Describe alternatives you've considered**
I forked this repository and implemented the change. I will open a pull request, but this would be my first real contribution to any open source project, so if I made some mistakes or my change does not make sense for you, feel free to decline the pr.

**Additional context**
When no `redisUrl` is set in the `credentials.json` file nothing should change, otherwise use the given config

### PR DESCRIPTION
**Pull request by nstringham**
Changes:
- only run app.js on non-master threads
- only run scripts on master thread
- only make one MongoDB instance per thread
- only make one Redis instance per thread
- allow redis URL to be configured and update Docs

closes #722 
closes #723

### CODE REVIEW

**Comment by nstringham** [comment]
@Flou21 can you make sure this works with your Redis setup

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -32,7 +32,8 @@ Before contributing to SkyCrypt, make sure you install the development environme
 3. On minecraft, log into `mc.hypixel.net`. Run the command `/api`, and copy the result.
 4. Open `credentials.json` and input your Hypixel API key into the `hypixel_api_key` field.
 5. In the `dbUrl` field, input your MongoDB url. In the `dbName` field, input the name of the database you would like to use.
-6. Making sure your Mongo and Redis instances are running, run `npm start` in the project directory. You should now be able to access the site at http://localhost:32464/
+6. (optional) If you are not using the default Redis port or you are using Redis remotely, you can configure the Redis URL with the `redisUrl` felid in `credentials.json`
```
Comment (dukio) [inline CONTRIBUTING.md:35]: Typo, "felid" instead of field

**Review by dukio — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (nstringham) [reply CONTRIBUTING.md]: Thanks

**Review by MartinNemi03 — APPROVED** [review]
Good job with this @nstringham.
LGTM, didn't test!

**Comment by Flou21** [comment]
> @Flou21 can you make sure this works with your Redis setup

yes it does solve my problem as well, thanks

File: src/app.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-const cluster = require("cluster");
+// this file never run on the master thread
```
Comment (metalcupcake5) [inline src/app.js:1]:
```suggestion
// this file never runs on the master thread
```
slight grammar adjustment :)

**Review by metalcupcake5 — APPROVED** [review]

**Pull request merged**